Initial Implementation of Event System

liamjwarren opened the pull request:
Initial Commit:
Completed
-Initial Keyboard event callbacks implemented.
-Initial Mouse event callbacks implemented.
-Logging completed for all callbacks, can be tested by building and information will appear for all callbacks.
-Set callbacks in Aquarius.cpp for Keyboard and Mouse completed.

TODO (in order)
-Implement Window event initial callbacks
-Set callbacks in Aquarius.cpp for Window.
-Expand event system
-Update logging messages to include key and mouse button names rather than their int representation.

colton-smith committed:
CMake: Add Aquarius/Src to include, and move glm include to after package search

colton-smith committed:
Add OpenGL rendering context initial implementation

colton-smith committed:
Add Window abstraction class (no callbacks registered)

colton-smith committed:
Update test main to utilise new Window class

colton-smith committed:
Remove RenderingContext destructor and shift responsibilites to Window destructor

liamjwarren committed:
Initial commit of event system

Completed
-Initial Keyboard event callbacks implemented.
-Initial Mouse event callbacks implemented.
-Logging completed for all callbacks, can be tested by building and information will appear for all callbacks.
-Set callbacks in Aquarius.cpp for Keyboard and Mouse completed.

TODO (in order)
-Implement Window event initial callbacks
-Set callbacks in Aquarius.cpp for Window.
-Expand event system
-Update logging messages to include key and mouse button names rather than their int representation.

liamjwarren committed:
Initial commit of event system

Completed
-Initial Keyboard event callbacks implemented.
-Initial Mouse event callbacks implemented.
-Logging completed for all callbacks, can be tested by building and information will appear for all callbacks.
-Set callbacks in Aquarius.cpp for Keyboard and Mouse completed.

TODO (in order)
-Implement Window event initial callbacks
-Set callbacks in Aquarius.cpp for Window.
-Expand event system
-Update logging messages to include key and mouse button names rather than their int representation.

liamjwarren committed:
Added new line at EOF

liamjwarren committed:
Added new line at EOF

Burke-Daniel commented on Aquarius/Src/Aquarius/Events/Keyboard.h:
```diff
@@ -0,0 +1,32 @@
+    public:
+        static void Callback(GLFWwindow* window, int key, int scanCode, int action, int modifications)
+        {
+            //TODO: Get it so key button name is printed rather then number, this is not critical but would be
+            // nice to have.
```
Not sure if you thought through how to accomplish this or if you have seen anyone else do it, but a `static const std::unordered_map<int, std::string>` comes to mind that you could throw in here as a lookup table

Burke-Daniel commented on Aquarius/Aquarius.cpp:
```diff
@@ -27,6 +30,16 @@ int Test::testMain()
     AQ_INFO("Window created successfully!");
 
+    //TODO Bring in neccesary set methods for window events.
+    glfwSetKeyCallback(window, KeyboardEvent::Callback);
+    glfwSetCursorPosCallback(window, MouseEvent::cursorPositionCallback);
+    glfwSetCursorEnterCallback(window, MouseEvent::cursorInWindowCallback);
+    glfwSetMouseButtonCallback(window, MouseEvent::mouseButtonCallback);
+    glfwSetScrollCallback(window, MouseEvent::mouseScrollCallback);
+
+    glfwSetInputMode(window, GLFW_STICKY_MOUSE_BUTTONS, 1);
+    glfwSetInputMode(window, GLFW_STICKY_KEYS,1);
```
All this should eventually happen within some of our internal code, either in the application or window initialization I would say

liamjwarren replied:
Good call, we can discuss the best spot for this.

Burke-Daniel commented on Aquarius/Src/Aquarius/Events/Keyboard.h:
```diff
@@ -0,0 +1,32 @@
+                    break;
+
+                case GLFW_REPEAT:
+                    AQ_INFO("key %v is being pressed repeatedly. ", key);
```
Extra space here at the end of the string

Burke-Daniel commented on Aquarius/Src/Aquarius/Events/Mouse.h:
```diff
@@ -0,0 +1,55 @@
+
+        static void cursorInWindowCallback(GLFWwindow* window, int inWindow)
+        {
+            if (inWindow)
```
Seeing an integer interpreted as a boolean is slightly misleading, but this is a consequence of glfw being written in C (C doesn't have a bool type), so this is fine here

Burke-Daniel commented on Aquarius/Src/Aquarius/Events/Mouse.h:
```diff
@@ -0,0 +1,55 @@
+        {
+            switch (action)
+            {
+                //TODO: Get it so mouse button name is printed rather then number, this is not critical but would be
+                // nice to have.
```
See comment above about unordered_map

Burke-Daniel reviewed (approved):
Works as expected when I run it, see lots of logging based off of keypresses, releases, and holds, and moving my mouse around in the window. I know Colton mentioned the same line braces so I didn't comment on those.

Have you thought about how this is going to fit in to the event system? right now these classes are basically just wrapping the callbacks we have to set in glfw. In terms of passing this information around the system, it might be good to have a base `Event` abstract class that all these inherit from. We can talk about this on the call. Other than that this is GTM after those couple of small style things

colton-smith committed:
Move GLFW error callback before GLFW init

liamjwarren committed:
Minor style corrections

- New line for braces beside namespace
- Addition of //namespace at end of all namespaces
- Removal of unintentional space in a string

liamjwarren committed:
Minor style corrections

- New line for braces beside namespace
- Addition of //namespace at end of all namespaces
- Removal of unintentional space in a string

colton-smith committed:
Move call to AQ_CORE_FATAL to avoid early program exit

colton-smith committed:
Change name of getVsync to isVsyncEnabled

colton-smith committed:
Fix Vsync disabled swap interval and reduce line length in Window constructor

colton-smith committed:
Merge pull request #16 from Burke-Daniel/feature/RE/window-and-context

Windowing and Rendering Context

liamjwarren committed:
Rebased from PR16

-Moved registering window callbacks to Window.cpp after rebasing from PR16.

liamjwarren committed:
Merge branch 'implement-event-system' of https://github.com/Burke-Daniel/Aquarius into implement-event-system

liamjwarren committed:
Revert "Rebased from PR16"

This reverts commit 158f6e131b906634d4589f34d2a1c7ff569d9d1b.

liamjwarren committed:
Revert "Merge branch 'implement-event-system' of https://github.com/Burke-Daniel/Aquarius into implement-event-system"

This reverts commit 965b89c760b3ea9fa869e6dbaef2a587b6171637, reversing
changes made to 158f6e131b906634d4589f34d2a1c7ff569d9d1b.

liamjwarren committed:
Revert "Revert "Rebased from PR16""

This reverts commit 09cd685bd653cb0fec6e0d5e8124519d8dbd97d6.
